Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -246,6 +246,35 @@ public final class SourceKitServer: LanguageServer {
 
     return nil
   }
+  
+  /// After the language service has crashed, send `DidOpenTextDocumentNotification`s to a newly instantiated language service for previously open documents.
+  func reopenDocuments(for languageService: ToolchainLanguageServer) {
+    queue.async {
+      guard let workspace = self.workspace else {
+        return
+      }
+      
+      for documentUri in workspace.documentManager.openDocuments where workspace.documentService[documentUri] === languageService {
+        guard let snapshot = workspace.documentManager.latestSnapshot(documentUri) else {
+          // The document has been closed since we retrieved its URI. We don't care about it anymore.
+          continue
+        }
+
+        // Close the docuemnt properly in the document manager and build system manager to start with a clean sheet when re-opening it.
+        let closeNotification = DidCloseTextDocumentNotification(textDocument: TextDocumentIdentifier(documentUri))
+        self.closeDocument(closeNotification, workspace: workspace)
+
+        let textDocument = TextDocumentItem(uri: documentUri,
+                                            language: snapshot.document.language,
+                                            version: snapshot.version,
+                                            text: snapshot.text)
+        // Re-open the document but don't register for build system notifications again since
+        // the document is still open in the build system server from before the sourcektid crash.
```

Comment:
This comment is now out of date. Otherwise, LGTM.

Reply:
Yikes. I’ve removed the comment. Let’s get this merged. 🚀